openshift/assisted-service
# MGMT-8789: Improve UX setting PUBLIC_CONTAINER_REGISTRIES

[MGMT-8789](https://issues.redhat.com/browse/MGMT-8789)
Previously, the only way for a user to add to `PUBLIC_CONTAINER_REGISTRIES` was by adding an annotation containing the configmap with the environment variable override [doc](https://github.com/openshift/assisted-service/blob/84d7256be1c24f1ec4dc1dadf02bbc775d5ff3d3/docs/operator.md#specifying-environmental-variables-via-configmap). The user experience of this is not ideal.

This change:
1. Adds a new spec field `unauthenticatedRegistries` to the `AgentServiceConfig`
2. Appends any registries specified in `unauthenticatedRegistries` and/or in the mirror registries to the default public container registry list 
    - Default list: quay.io, registry.svc.ci.openshift.org

Any registries listed in `PUBLIC_CONTAINER_REGISTRIES` will automatically be ignored for pull secret validation in assisted service, meaning these registries do not require an entry in the user's pull secret.



## List all the issues related to this PR

https://issues.redhat.com/browse/MGMT-8789


## What environments does this code impact?


## How was this code tested?




Testing:
1. Deployed an OCP cluster and ran the operator with these changes
````
$ operator-sdk run bundle --namespace assisted-installer ${BUNDLE_IMAGE:-quay.io/edge-infrastructure/assisted-service-operator-bundle:latest} --kubeconfig ${KUBECONFIG} --pull-secret-name ${OPERATOR_PULL_SECRET}
````
3. Created an `AgentServiceConfig` with a mirror registry configured with `example.com` present
4. Set the `OPENSHIFT_INSTALL_RELEASE_IMAGE_MIRROR` environment variable for the assisted-service to `example.com/test/registry:tag-1` 
    - Service will check the pull secret for a registry entry of `example.com`
5. Create a k8s pull secret without `example.com`
6. Create an InfraEnv referencing the pull secret
7. Check the logs of assisted-service to ensure the InfraEnv passes pull secret validation
8. Edit the `AgentServiceConfig` and remove the mirror registry ref
9. Check the logs of assisted-service to see the InfraEnv fail pull secret validation
10. Edit the `AgentServiceConfig` and add `example.com` and `quay.io` to the `unauthenticatedRegistries` field
7. Check the logs of assisted-service to ensure the InfraEnv passes pull secret validation
8. Check the assisted service configmap to ensure the `PUBLIC_CONTAINER_REGISTRIES` was set correctly
````
$ oc get configmap -n assisted-installer assisted-service -oyaml | grep PUBLIC
````

## Checklist


## Reviewers Checklist

- Are the title and description (in both PR and commit) meaningful and clear?
- Is there a bug required (and linked) for this change?
- Should this PR be backported?

[Kubernetes community documentation]: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#commit-message-guidelines
[CONTRIBUTING]: https://github.com/openshift/assisted-service/blob/master/CONTRIBUTING.md

/cc @carbonin @filanov @flavio-fernandes
